File: 2-dead-letter-and-circuit-breaker/0-typescript/src/api.controller.ts
import { BadRequestException, Body, Controller, Get, Param, Post, Query } from "@nestjs/common"
import { InjectRepository } from "@nestjs/typeorm"
import { Repository } from "typeorm"
import { CircuitService, FAIL_THRESHOLD, OPEN_TTL_SEC } from "./circuit.service"
import { DeadLetter, OutboxEvent } from "./entities"

interface PublishDto {
    type: string
    payload: Record<string, unknown>
    targetUrl: string
}

@Controller("api")
export class ApiController {
    constructor(
        @InjectRepository(OutboxEvent) private readonly outboxRepo: Repository<OutboxEvent>,
        @InjectRepository(DeadLetter) private readonly dlqRepo: Repository<DeadLetter>,
        private readonly circuit: CircuitService,
    ) {}

    /** Publish writes one durable pending outbox row; the poller takes it from there. */
    @Post("events/publish")
    async publish(@Body() dto: PublishDto): Promise<{ eventId: string; status: string }> {
        const row = await this.outboxRepo.save(
            this.outboxRepo.create({
                type: dto.type,
                payload: dto.payload,
                targetUrl: dto.targetUrl,
                status: "pending",
            }),
        )
        return { eventId: row.id, status: "pending" }
    }

    /** List parked dead-letter rows, newest first. */
    @Get("dlq")
    async listDlq(): Promise<DeadLetter[]> {
        return this.dlqRepo.find({ order: { parkedAt: "DESC" } })
    }

    /**
     * Manual replay: stamp replayedAt, then flip the original outbox row back to
     * pending so the poller re-enqueues it. The DLQ row stays for audit.
     */
    @Post("dlq/:id/replay")
    async replay(
        @Param("id") id: string,
    ): Promise<{ replayedDlq: string; eventId: string; status: string; note: string }> {
        const row = await this.dlqRepo.findOne({ where: { id } })
        if (!row) {
            throw new BadRequestException("no such DLQ row")
        }
        await this.dlqRepo.update({ id }, { replayedAt: new Date() })
        await this.outboxRepo.update(
            { id: row.eventId },
            { status: "pending", attempts: 0, lastError: null },
        )
        return {
            replayedDlq: id,
            eventId: row.eventId,
            status: "requeued",
            note: "Outbox row reset to pending; poller will re-enqueue within ~500ms.",
        }
    }

    /** Inspect the circuit state for one host. */
    @Get("circuit/status")
    async circuitStatus(@Query("host") host: string): Promise<{
        host: string
        state: string
        failCount: number
        failThreshold: number
        openTtlSec: number
    }> {
        return {
            host,
            state: await this.circuit.getState(host),
            failCount: await this.circuit.failCount(host),
            failThreshold: FAIL_THRESHOLD,
            openTtlSec: OPEN_TTL_SEC,
        }
    }
}

File: 2-dead-letter-and-circuit-breaker/1-java/src/main/java/com/starci/webhook/ApiController.java
package com.starci.webhook;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.net.URI;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api")
public class ApiController {
    private final OutboxRepository outbox;
    private final DeadLetterRepository dlq;
    private final CircuitService circuit;

    ApiController(OutboxRepository outbox, DeadLetterRepository dlq, CircuitService circuit) {
        this.outbox = outbox;
        this.dlq = dlq;
        this.circuit = circuit;
    }

    @PostMapping("/events/publish")
    public Map<String, Object> publish(@RequestBody PublishDto dto) {
        OutboxEvent ev = new OutboxEvent();
        ev.type = dto.type();
        ev.payload = dto.payload();
        ev.targetUrl = dto.targetUrl();
        ev.status = "pending";
        outbox.save(ev);
        return Map.of("eventId", ev.id.toString(), "status", "pending");
    }

    @GetMapping("/dlq")
    public List<DeadLetter> listDlq() {
        return dlq.findAllByOrderByParkedAtDesc();
    }

    @PostMapping("/dlq/{id}/replay")
    public Map<String, Object> replay(@PathVariable UUID id) {
        return dlq.findById(id).map(row -> {
            row.replayedAt = OffsetDateTime.now();
            dlq.save(row);
            outbox.findById(row.eventId).ifPresent(ev -> {
                ev.status = "pending";
                ev.attempts = 0;
                ev.lastError = null;
                outbox.save(ev);
            });
            return Map.<String, Object>of(
                    "replayedDlq", id.toString(),
                    "eventId", row.eventId.toString(),
                    "status", "requeued",
                    "note", "Outbox row reset to pending; poller will re-enqueue within ~500ms.");
        }).orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "no such DLQ row"));
    }

    @GetMapping("/circuit/status")
    public Map<String, Object> circuitStatus(@RequestParam String host) {
        long ttl = circuit.openTtlMillis(host);
        return Map.of(
                "host", host,
                "state", circuit.getState(host),
                "failCount", circuit.failCount(host),
                "failThreshold", CircuitService.FAIL_THRESHOLD,
                "openTtlSec", ttl > 0 ? (int) (ttl / 1000) : 0);
    }

    record PublishDto(String type, Map<String, Object> payload, String targetUrl) {}

    @SuppressWarnings("unused")
    private static String hostOf(String url) {
        URI u = URI.create(url);
        return u.getHost();
    }
}
